File: macos/QMK Toolbox/HIDConsole/HIDConsoleDevice.m
#import "HIDConsoleDevice.h"

@implementation HIDConsoleDevice {
    uint8_t hidReportBuffer[64];
}

- (id)initWithDeviceRef:(IOHIDDeviceRef)deviceRef {
    if (self = [super init]) {
        self.deviceRef = deviceRef;

        self.manufacturerString = [self stringProperty:CFSTR(kIOHIDManufacturerKey)];
        self.productString = [self stringProperty:CFSTR(kIOHIDProductKey)];
        self.vendorID = [self ushortProperty:CFSTR(kIOHIDVendorIDKey)];
        self.productID = [self ushortProperty:CFSTR(kIOHIDProductIDKey)];
        self.revisionBCD = [self ushortProperty:CFSTR(kIOHIDVersionNumberKey)];
        self.currentLine = @"";

        IOHIDDeviceRegisterInputReportCallback(self.deviceRef, hidReportBuffer, sizeof(hidReportBuffer), reportReceived, (__bridge void *)self);
    }
    return self;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ %@ (%04X:%04X:%04X)", self.manufacturerString, self.productString, self.vendorID, self.productID, self.revisionBCD];
}

static void reportReceived(void *context, IOReturn result, void *sender, IOHIDReportType type, uint32_t reportID, uint8_t *report, CFIndex reportLength) {
    HIDConsoleDevice *const device = (__bridge HIDConsoleDevice *const)context;

    // Check if we have a completed line queued
    NSUInteger lineEnd = [device.currentLine rangeOfString:@"\n"].location;
    if (lineEnd == NSNotFound) {
        // Partial line or nothing - append incoming report to current line
        NSString *reportString = [NSString stringWithCString:(char *)report encoding:NSUTF8StringEncoding];
        device.currentLine = [device.currentLine stringByAppendingString:reportString];
    }

    // Check again for a completed line
    lineEnd = [device.currentLine rangeOfString:@"\n"].location;
    while (lineEnd != NSNotFound) {
        // Fire delegate with completed lines until we have none left
        NSString *completedLine = [device.currentLine substringToIndex:lineEnd];
        device.currentLine = [device.currentLine substringFromIndex:lineEnd + 1];
        lineEnd = [device.currentLine rangeOfString:@"\n"].location;
        [device.delegate consoleDevice:device didReceiveReport:completedLine];
    }
}

- (NSString *)stringProperty:(CFStringRef)property {
    return (__bridge NSString *)IOHIDDeviceGetProperty(self.deviceRef, property);
}

- (ushort)ushortProperty:(CFStringRef)property {
    return [(__bridge NSNumber *)IOHIDDeviceGetProperty(self.deviceRef, property) unsignedShortValue];
}

- (nonnull id)copyWithZone:(nullable NSZone *)zone {
    HIDConsoleDevice *deviceCopy = [[[self class] allocWithZone:zone] init];

    if (deviceCopy) {
        deviceCopy.deviceRef = self.deviceRef;
        deviceCopy.manufacturerString = self.manufacturerString;
        deviceCopy.productString = self.productString;
        deviceCopy.vendorID = self.vendorID;
        deviceCopy.productID = self.productID;
        deviceCopy.revisionBCD = self.revisionBCD;
    }

    return deviceCopy;
}
@end

File: macos/QMK Toolbox/HIDConsole/HIDConsoleDevice.swift
import Foundation

protocol HIDConsoleDeviceDelegate: AnyObject {
    func consoleDevice(_ device: HIDConsoleDevice, didReceiveReport report: String)
}

class HIDConsoleDevice: Equatable, CustomStringConvertible {
    weak var delegate: HIDConsoleDeviceDelegate?

    private var reportBuffer: UnsafeMutablePointer<UInt8>

    private var reportBufferSize: Int = 0

    let hidDevice: IOHIDDevice

    var manufacturer: String? {
        HIDConsoleDevice.stringProperty(kIOHIDManufacturerKey, for: hidDevice)
    }

    var product: String? {
        HIDConsoleDevice.stringProperty(kIOHIDProductKey, for: hidDevice)
    }

    var vendorID: UInt16 {
        HIDConsoleDevice.uint16Property(kIOHIDVendorIDKey, for: hidDevice)
    }

    var productID: UInt16 {
        HIDConsoleDevice.uint16Property(kIOHIDProductIDKey, for: hidDevice)
    }

    var revisionBCD: UInt16 {
        HIDConsoleDevice.uint16Property(kIOHIDVersionNumberKey, for: hidDevice)
    }

    init(_ device: IOHIDDevice) {
        hidDevice = device
        reportBufferSize = IOHIDDeviceGetProperty(device, kIOHIDMaxInputReportSizeKey as CFString) as! Int
        reportBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: reportBufferSize)

        let inputReportCallback: IOHIDReportCallback = { context, result, sender, type, reportID, report, length in
            let device = Unmanaged<HIDConsoleDevice>.fromOpaque(context!).takeUnretainedValue()
            let reportData = Data(bytes: report, count: device.reportBufferSize)
            device.reportReceived(reportData)
        }

        let unsafeSelf = Unmanaged.passRetained(self).toOpaque()
        IOHIDDeviceRegisterInputReportCallback(hidDevice, reportBuffer, reportBufferSize, inputReportCallback, unsafeSelf)
    }

    private var currentLine = Data()

    func reportReceived(_ report: Data) {
        // Check if we have a completed line queued
        var lineEnd = currentLine.firstIndex(of: UInt8(ascii: "\n"))
        if lineEnd == nil {
            // Partial line or nothing - append incoming report to current line
            for b in report {
                // Trim trailing null bytes
                if b == 0 {
                    break
                }
                currentLine.append(b)
            }
        }

        // Check again for a completed line
        lineEnd = currentLine.firstIndex(of: UInt8(ascii: "\n"))
        while lineEnd != nil {
            // Fire delegate with completed lines until we have none left
            // Only convert to string at the last possible moment in case there is a UTF-8 sequence split across reports
            guard let completedLine = String(data: currentLine[..<lineEnd!], encoding: .utf8) else { return }
            currentLine = currentLine[currentLine.index(after: lineEnd!)...]
            lineEnd = currentLine.firstIndex(of: UInt8(ascii: "\n"))
            delegate?.consoleDevice(self, didReceiveReport: completedLine)
        }
    }

    var description: String {
        String(format: "%@ %@ (%04X:%04X:%04X)", manufacturer ?? "", product ?? "", vendorID, productID, revisionBCD)
    }

    static func == (lhs: HIDConsoleDevice, rhs: HIDConsoleDevice) -> Bool {
        return lhs.hidDevice === rhs.hidDevice
    }

    static func stringProperty(_ propertyName: String, for device: IOHIDDevice) -> String? {
        return IOHIDDeviceGetProperty(device, propertyName as CFString) as! String?
    }

    static func uint16Property(_ propertyName: String, for device: IOHIDDevice) -> UInt16 {
        return (IOHIDDeviceGetProperty(device, propertyName as CFString) as! NSNumber?)!.uint16Value
    }
}
